design-documents/storefront/catalog/product-options-and-variants.md

```diff
@@ -0,0 +1,363 @@
+and is this option mandatory.  
+* **ProductOptionValue**s - belong to a particular option as to a group and represent selections which allowed by the option.
+Option value could by display label, also one or may values could be pre-selected.  
+* **ProductVariant** - represent the final selection of one or multiple option values that will characterize the customized product in a shopping cart.
+Depends on the business scenario, a particular product variant could be linked with an existing product,
+with a price, or an inventory record, or no be linked to any.
+Even with no entities associated with the variant, it still has great value for a catalog because the presence of the variant says that such composition of options and their values described by the variant makes sense so that it can be purchased.
+   
+### Actual usages
+
+The application distinguishes two approaches to manage options:
+
+**Approach 1, "One to many"**: Multiple options selections lead a shopper to a selection of the single variation.
+
+Example: configurable products
+![](https://app.lucidchart.com/publicSegments/view/26cd0b67-13c1-44e4-8b61-cada36c67010/image.png)
+
+**Approach 2, "One to one"**: The selection of multiple options leads to multiple product variants.
+
+Examples: bundle product, customizable product, downloadable products.
+![](https://app.lucidchart.com/publicSegments/view/fea3f950-6f2f-4e46-90e4-f8ee4b6877f7/image.png)
+Both of approaches could be used together.
+
+Example: configurable product with customizable option.
+
+
+## Competitors analysis
+During my research, I analyzed documentation on how options are covered by the competitors.
+commercetools and Shopify attack this domain with product variants that represent all the possible options.
+Despite the domain simplification by reducing the number of entities both systems constrained with the size of a product which makes it impossible to compose complex business cases that we frequently may face in the Magento ecosystem.
+
+[Commercetools: Modeling Products](https://docs.commercetools.com/product-modeling-products)
+
+[Shopify: Variants](https://help.shopify.com/en/manual/products/variants)
+
+Bigcommerce has a more complex option representation then Shopify or Commercetools.
+in some way, I would compare their options features with Magento 2 x functionalities.
+Bigcommerce segregates options from variants and allows to associate intersection of options with a new product, and as a result, I believe, struggles from the same disease as Magento does - redundancy in the variant data.
+
+[Bigcommerce: Product Options](https://support.bigcommerce.com/s/article/Options-SKUs-Rules)
+
+## Goals
+
+* Eliminate hard dependency on prices from options to make possible use promotions and B2B prices for customizable options, bundle, and downloadable products.
+* Eliminate variants redundancy
+    * Catalog should not have a variant per each option values combination, only if necessary.
```

Comment:
this is exactly what commercetoool is doing, so I would agree with this approach. Also they are not using Options concept at all, rather Product Type and Attributes (with variant constraints).